airbyte-integrations/connectors/source-google-ads/source_google_ads/google_ads.py

```diff
@@ -52,48 +58,51 @@ def send_request(self, query: str) -> SearchGoogleAdsResponse:
     @staticmethod
     def get_fields_from_schema(schema: Mapping[str, Any]) -> List[str]:
         properties = schema.get("properties")
-
         return [*properties]
 
     @staticmethod
-    def convert_schema_into_query(schema: Mapping[str, Any], report_name: str, from_date: str, to_date: str) -> str:
+    def convert_schema_into_query(
+        schema: Mapping[str, Any], report_name: str, from_date: str = None, to_date: str = None, cursor_field: str = None
+    ) -> str:
         from_category = REPORT_MAPPING[report_name]
         fields = GoogleAds.get_fields_from_schema(schema)
         fields = ",\n".join(fields)
 
-        query = Template(
-            """
-          SELECT
-            $fields
-          FROM $from_category
-          WHERE segments.date > '$from_date'
-            AND segments.date < '$to_date'
-          ORDER BY segments.date
-      """
-        )
-        query = query.substitute(fields=fields, from_category=from_category, from_date=from_date, to_date=to_date)
+        query_template = "SELECT {fields} FROM {from_category} "
+        substitute_params = dict(fields=fields, from_category=from_category)
+
+        if cursor_field:
+            query_template += "WHERE {cursor_field} > '{from_date}' AND {cursor_field} < '{to_date}' ORDER BY {cursor_field}"
+            substitute_params.update(dict(from_date=from_date, to_date=to_date, cursor_field=cursor_field))
 
-        return query
+        return query_template.format(**substitute_params)
 
     @staticmethod
-    def get_field_value(result: GoogleAdsRow, field: str) -> str:
+    def get_field_value(field_value: GoogleAdsRow, field: str) -> str:
         field_name = field.split(".")
-        try:
-            field_value = result
-            for level_attr in field_name:
-                field_value = field_value.__getattr__(level_attr)
-                if isinstance(field_value, Enum):
-                    field_value = field_value.name
+        for level_attr in field_name:
+            try:
+                field_value = getattr(field_value, level_attr)
+            except AttributeError:
+                field_value = getattr(field_value, level_attr + "_", None)
+
+            if isinstance(field_value, Enum):
+                field_value = field_value.name
+            elif isinstance(field_value, (Repeated, RepeatedComposite)):
+                field_value = [str(value) for value in field_value]
+
+        # Google Ads has a lot of entities inside itself and we cannot process them all separately, because:
+        # 1. It will take a long time
+        # 2. We have no way to get data on absolutely all entities to test.
+        #
+        # To prevent JSON from throwing an error during deserialization, we made such a hack.
+        if not (isinstance(field_value, (list, int, float, str, bool, dict)) or field_value is None):
```

Comment:
this is interesting - what's an example of a value that is not an instanceof these data types? can you include a reference to one such field? I'm trying to understand why exactly this failure might happen

Reply:
Yup, for example:
1. [ad_group_ad.ad.responsive_display_ad.long_headline](https://developers.google.com/google-ads/api/fields/v8/ad_group_ad#ad_group_ad.ad.responsive_display_ad.long_headline) - This value is of type [AdTextAsset](https://developers.google.com/google-ads/api/reference/rpc/v6/AdTextAsset?hl=en).
2. [ad_group_ad.ad.legacy_app_install_ad](https://developers.google.com/google-ads/api/fields/v8/ad_group_ad#ad_group_ad.ad.legacy_app_install_ad) - [LegacyAppInstallAdInfo](https://developers.google.com/google-ads/api/reference/rpc/v7/LegacyAppInstallAdInfo?hl=en)

And there are a lot of such different types, and many more. The previous author of this code just wrapped each value in a string and had no problems. I tried to get values of the correct type, but sometimes these are additional nested objects, sometimes arrays. And JSON cannot convert an object of some Python class into its own object.

Reply:
could you add these examples as comments in the code?